src/hyperlink/_url.py

```diff
@@ -1991,9 +2005,20 @@ class DecodedURL(object):
         lazy (bool): Set to True to avoid pre-decode all parts of the URL to
             check for validity. Defaults to False.
 
+    .. note::
+
+      The :class:`DecodedURL` initializer takes a :class:`URL` object,
+      not URL components, like :class:`URL`. To programmatically
+      construct a :class:`DecodedURL`, you can use this pattern:
+
+        >>> print(DecodedURL().replace(scheme=u'https',
+        ... host=u'pypi.org', path=(u'projects', u'hyperlink')).to_text())
+        https://pypi.org/projects/hyperlink
+
+    .. versionadded:: 18.0.0
     """
 
-    def __init__(self, url, lazy=False):
+    def __init__(self, url=_EMPTY_URL, lazy=False):
```

Comment:
why the default? Perhaps merits discussion outside of doc changes.

Reply:
Sure, a couple reasons:

1. It's more parallel with `URL()`, which can be instantiated without arguments to get an empty URL.
2. This way to programmatically construct a `DecodedURL`, you don't need to also import `URL`. You can do `DecodedURL().replace(...)`.

Reply:
Ah, and the reason for including it in these changes was because I really wanted to avoid people programmatically constructing the newly-exposed DecodedURL by doing

```
from hyperlink import URL, DecodedURL
DecodedURL(URL(...))
```
Without realizing that URL's initializer arguments aren't as rigorously decoded/encoded as the rest of DecodedURL's API.

I proposed adding a `.build()` or [`.from_parts()`](https://github.com/python-hyper/hyperlink/issues/11#issuecomment-370124799) if we want to shift to a better pattern. For now it's either that or [`parse('').replace(...)`](https://github.com/python-hyper/hyperlink/issues/11#issuecomment-370096947). I figured a default arg at least lets you explicitly use the type without an extra `parse()`/`from_text()` step.

Reply:
`from_parts()` sounds like a good addition. I have written `DecodedURL(URL())` &mdash; it's a bit awkward. Perhaps I am worrying too much (this is Python, everything is slow...) but I tend to look for APIs that avoid temporaries.

Reply:
OK makes sense, and yes to `from_parts`.